apps/emqx_message_transformation/src/emqx_message_transformation_http_api.erl

```diff
@@ -305,15 +306,29 @@ fields(dryrun_input_message) ->
     %% See `emqx_message_transformation:eval_context()'.
     [
         {client_attrs, mk(map(), #{default => #{}})},
+        {clientid, mk(binary(), #{default => <<"test-clientid">>})},
+        {flags, mk(ref(dryrun_input_flags), #{default => #{}})},
         {payload, mk(binary(), #{required => true})},
+        {peername, mk(emqx_schema:ip_port(), #{default => <<"127.0.0.1:19872">>})},
+        {pub_props,
+            mk(
+                typerefl:alias("map()", publish_properties()),
+                #{default => #{}}
+            )},
         {qos, mk(range(0, 2), #{default => 0})},
         {retain, mk(boolean(), #{default => false})},
         {topic, mk(binary(), #{required => true})},
         {user_property,
             mk(
                 typerefl:alias("map(binary(), binary())", user_property()),
                 #{default => #{}}
-            )}
+            )},
+        {username, mk(binary(), #{required => false})}
+    ];
+fields(dryrun_input_flags) ->
+    [
+        {dup, mk(boolean(), #{default => false})},
```

Comment:
I don't think any message should be explicit set dup flag to 'true'.
dup is only set when a message is re-transmitted after session resume.

Reply:
This is for dry run testing a transformation, so it would make sense to allow simulating a duplicate message?

Reply:
the dup flag might be set by the publish-client, but it's cleared before calling the message.publish hook
https://github.com/emqx/emqx/blob/acb8c994fa4825af988f36fa6929853e0204a21d/apps/emqx/src/emqx_broker.erl#L246

the dup flag is only set before re-sending a message to subscriber-client
https://github.com/emqx/emqx/blob/master/apps/emqx/src/emqx_session_mem.erl#L661

Reply:
I see, thanks for the clarification.  Removed this.